Podcasting: add upgrade nudge to podcast settings page

creativecoder opened the pull request:
## Summary

This PR adds an upgrade nudge to the top of the podcasting details page for sites with a free plan. It also re-enables the `plan` property on the upgrade nudge, so that we can specify which plan to highlight on the plans page.

## Details

As part of the PR, we update how features are described for each plan in `constants/plans`
1. Rename `getFeatures` to `getPlanCompareFeatures`, to clarify that this list of features is not complete, but mainly used to populate the `PlanFeatures` component when comparing plans.
1. Adds `getHiddenFeatures` to include features that are part of plan abilities but not shown to users.

Additionally `planHasFeature` in `client/lib/plans` is refactored to check against all features in a plan, not just those listed on the plans comparison page.

These changes were needed because
1. There was no single feature currently displayed that we could test for related to audio uploads. (Space upgrades are shown, but those are individual features tied to amount of space, and each plan has a different one)
1. Even though audio uploads are included with all paid plans, this feature is not displayed on the plans page, so initially we were not able to use the feature to determine if the upgrade nudge should display.

## Screenshot

<img width="759" alt="podcast settings upgrade nudge" src="https://user-images.githubusercontent.com/1699996/43915448-55d682d2-9bd0-11e8-9c7f-29447489a513.png">

## Testing instructions

Tests that review the plans pages are also included, because this PR modifies that code.

#### Upgrade Nudge on Podcast Settings Page

1. On a site with a free plan, go to the podcast settings (Settings > Writing > Podcasting).
1. Verify that the upgrade nudge is shown above the settings card with messaging for Personal plan and audio uploads with no console errors.
1. Verify that clicking on the upgrade nudge takes you to the plans page with the Personal plan "suggested".
1. Verify that all features display correctly on the plans page and there are no console errors.
1. On a site with a paid plan (Personal, Premium, or Business) go to the podcast settings (Settings > Writing > Podcasting).
1. Verify that the upgrade nudge is _not_ shown at the top of page.

#### Jetpack Plans

1. With a site that is not connected to Jetpack, connect that site to Jetpack (spin up a new site with jurassic.ninja, if needed).
1. After you click the "Approve" button to connect the site to WordPress.com, you should be on the Jetpack plans page (but on production wordpress.com).
1. Reload the Jetpack plans page with Calypso dev url so you are viewing code in this PR (`http://calypso.localhost:3000/jetpack/connect/plans/{site}`).
1. Verify that the plans features display correctly on the Jetpack plans page and there are no console errors.
1. Purchase a plan and check out.
1. After you check out, you should see a Jetpack "Thank You" page.
1. Verify that the plan features display correctly on the thank you page and there are no console errors.

#### Podcast settings for Jetpack sites 

1. Visit the podcast settings page for the Jetpack site (`/settings/podcasting/{site}`).
2. You should see an error about not being able to manage podcast settings for the site, and the upgrade nudge should not be shown, with no console errors.

mattsherman commented:
Taking a look at this.

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -184,7 +184,17 @@ const getPlanPersonalDetails = () => ( {
+			...getPlanPersonalDetails().getDefaultPlanListFeatures(),
+			...getPlanPersonalDetails().getSignupFeatures(),
+			...getPlanPersonalDetails().getBlogSignupFeatures(),
+			...getPlanPersonalDetails().getPortfolioSignupFeatures(),
```
Can you use `...this.getDefaultPlanListFeatures()` instead? That would be more readable than this.

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -232,7 +242,18 @@ const getPlanPremiumDetails = () => ( {
+			...getPlanPremiumDetails().getPromotedFeatures(),
+			...getPlanPremiumDetails().getSignupFeatures(),
+			...getPlanPremiumDetails().getBlogSignupFeatures(),
+			...getPlanPremiumDetails().getPortfolioSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -312,7 +333,18 @@ const getPlanBusinessDetails = () => ( {
+			...getPlanBusinessDetails().getPromotedFeatures(),
+			...getPlanBusinessDetails().getSignupFeatures(),
+			...getPlanBusinessDetails().getBlogSignupFeatures(),
+			...getPlanBusinessDetails().getPortfolioSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -378,7 +410,15 @@ export const PLANS_LIST = {
+				...PLANS_LIST[ PLAN_FREE ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_FREE ].getSignupFeatures(),
+				...PLANS_LIST[ PLAN_FREE ].getBlogSignupFeatures(),
+				...PLANS_LIST[ PLAN_FREE ].getPortfolioSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -498,7 +538,13 @@ export const PLANS_LIST = {
+		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_FREE ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_FREE ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -545,7 +591,13 @@ export const PLANS_LIST = {
 		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_PREMIUM ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_PREMIUM ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -597,7 +649,13 @@ export const PLANS_LIST = {
 		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_PREMIUM_MONTHLY ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_PREMIUM_MONTHLY ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -648,7 +706,13 @@ export const PLANS_LIST = {
+		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_PERSONAL ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_PERSONAL ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -687,7 +751,13 @@ export const PLANS_LIST = {
+		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_PERSONAL_MONTHLY ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_PERSONAL_MONTHLY ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -737,7 +807,13 @@ export const PLANS_LIST = {
 		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_BUSINESS ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_BUSINESS ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

mattsherman commented on client/lib/plans/constants.js:
```diff
@@ -799,7 +875,13 @@ export const PLANS_LIST = {
 		getFeatures: () =>
+			uniq( [
+				...PLANS_LIST[ PLAN_JETPACK_BUSINESS_MONTHLY ].getDefaultPlanListFeatures(),
+				...PLANS_LIST[ PLAN_JETPACK_BUSINESS_MONTHLY ].getSignupFeatures(),
```
Same comment as above... can you use `this`?

creativecoder commented:
> Can you use ...this.getDefaultPlanListFeatures() instead? That would be more readable than this.

@mattsherman Good question! I had the same instinct at first to use `this` within all of the `getFeatures` methods, and we tried this in the first pass. However, the way `getFeatures` is called through several layers makes `this === undefined` in every context I've seen so far.

We could use `.call` or `.bind` to explicitly define `this` when `getFeatures` is called, but that would require lots of other reworking across the files where `getFeatures` is called.

> this file isn't just constants, and there seems to be a lot of code duplication

Agreed. This approach attempts to take a middle road by extending how plan features are defined for the needs of this PR, without a more invasive restructuring. But let me think on this a bit more... maybe a small helper function could improve this without causing other problems.

creativecoder commented:
@mattsherman Here's a snippet that would reduce duplication.

Add helper functions near the top of the file like this

```js
const reducePlanFeatures = ( planContext, featuresArray, featureMethod ) => {
	if ( planContext[ featureMethod ] && typeof planContext[ featureMethod ] === 'function' ) {
		return [ ...planContext[ featureMethod ](), ...featuresArray ];
	}

	return featuresArray;
};

const collectPlanFeatures = ( planContext, features = [] ) => {
	const allFeatures = [
		'getDefaultPlanListFeatures',
		'getPromotedFeatures',
		'getSignupFeatures',
		'getBlogSignupFeatures',
		'getPortfolioSignupFeatures',
	].reduce( reducePlanFeatures.bind( null, planContext ), features );

	return uniq( allFeatures );
};
```

Then for each instance of `getFeatures` throughout the file, do something like this instead

```js
getFeatures: () => collectPlanFeatures( getPlanPersonalDetails, [
	FEATURE_AUDIO_UPLOADS,
	FEATURE_VIDEO_UPLOADS,
] ),
```

Even though this reduces duplication, I think this way is harder to read, thus I'd vote for keeping the duplication as is, but being more explicit.

mattsherman commented:
> We could use .call or .bind to explicitly define this when getFeatures is called, but that would require lots of other reworking across the files where getFeatures is called.

Yeah, that doesn't sound ideal.

> This approach attempts to take a middle road by extending how plan features are defined for the needs of this PR, without a more invasive restructuring.

> I'd vote for keeping the duplication as is, but being more explicit.

Agreed. Thanks for thinking about this more. Let's keep the duplication for now.

dzver commented:
Let me try to explain what I understand from this diff and please tell me if I get it right or not.

This diff does two things:
1. It refactors the plan features, repurposing getFeatures from a function that returns features for presentation to a function that only serves hasFeature
2. It adds a nudge

Is that correct? 

In case it is, I have follow-up questions (that only apply if I got it right):
- Is that function used in other places for presentation that will now change?
- Is it reasonable to solve the problem by combining the features in hasFeature thus making a smaller refactoring?

creativecoder commented:
@dzver Thanks for taking a look!

> This diff does two things:
>
> 1. It refactors the plan features, repurposing getFeatures from a function that returns features for presentation to a function that only serves hasFeature
> 2. It adds a nudge

Both correct. Note that I've just renamed `getFeatures` to `getAllFeatures` for each plan to make the purpose a little clearer.

> Is that function used in other places for presentation that will now change?

Yes, you can see those places in the PR diff. They have been updated to use `getDefaultPlanListFeatures`

1. Within `applyTestFiltersToPlansList` in `client/lib/plans/index.js`
    - Function used to filter which plan features are displayed in a/b tests (https://github.com/Automattic/wp-calypso/pull/26575/files#diff-907fd275ac1e8e37aab2426a2e010b12R372)
    - Called on the plans page where plan features are compared (https://github.com/Automattic/wp-calypso/pull/26575/files#diff-801f5ceb927d2d46a3777a0f162507d7R596)
        <img width="1452" alt="plan page" src="https://user-images.githubusercontent.com/1699996/44119298-77a03946-9fde-11e8-8e2c-d01634e5d56e.png">
    - Also called on the manage purchase page, though features are not displayed here, only plan description
        <img width="1452" alt="manage purchase" src="https://user-images.githubusercontent.com/1699996/44119344-909bd25c-9fde-11e8-9a40-89448cb9d419.png">

2. Within the Jetpack Thank You card used after completing checkout with a Jetpack plan
    - https://github.com/Automattic/wp-calypso/pull/26575/files#diff-a256b914f277ea5af200ec47e3ffe7fdR648
    <img width="1452" alt="jetpack thank you card" src="https://user-images.githubusercontent.com/1699996/44119446-d18864b0-9fde-11e8-8cf4-8d5dee9c1c94.png">

> Is it reasonable to solve the problem by combining the features in hasFeature thus making a smaller refactoring?

I like that idea in principle -- let me try it out and I'll post an update here.

mattsherman commented:
@creativecoder Setting back to "In Progress" to match the latest comments. Please relabel it as "Needs Review" and move to the "Under Review" column on the work board when it is ready to be looked at again!

creativecoder commented on client/lib/plans/index.js:
```diff
@@ -69,8 +69,35 @@ export function getPlanPath( plan ) {
+		[]
+	);
+
+	return includes( allFeatures, feature );
```
One risk in this refactor is having this function return a different result if testing for features that are not in the `getFeatures/getPlanListFeatures` function, but are in other features functions (like `getPromotedFeatures`, `getSignupFeatures`, etc), since all of those additional functions are now included in the feature checking.

I did thorough checking of the `feature` argument at all calls of `planHasFeature` (both calling this function directly and through `hasFeature` selector in `client/state/sites/plans/selectors.js`). All calls use features that are listed in `getFeatures/getPlanListFeatures`, and thus the return value will be the same before and after this refactor.

creativecoder commented on client/lib/plans/constants.js:
```diff
@@ -184,7 +184,7 @@ const getPlanPersonalDetails = () => ( {
 			}
 		),
-	getFeatures: () => [
+	getPlanListFeatures: () => [
```
Renamed these to try and make it clearer that this is a subset of features for each plan, specifically used to populate the PlanList component.

creativecoder commented on client/state/sites/plans/selectors.js:
```diff
@@ -201,11 +200,6 @@ export function getSitePlanSlug( state, siteId ) {
-function planHasFeature( plan, feature ) {
-	return includes( get( PLANS_LIST[ plan ], 'getFeatures', () => [] )(), feature );
-}
-
```
Replaced with new implementation of `planHasFeature` from `client/lib/plans/index.js`, as was done in https://github.com/Automattic/wp-calypso/pull/9635.

creativecoder commented:
@mattsherman @dzver This is ready for re-review.

Note the following changes

1. Refactored to collect all plan features in `planHasFeature` to reduce changes needed in `plan/constants` (thanks for the suggestion @dzver!)
1. Hide upgrade nudge until plans data has loaded to prevent it showing for users with a plan while page is loading
1. Added Jetpack e2e tests to PR and Jetpack specific testing instructions
1. Added relevant unit tests for functionality in this PR
1. Revised PR description with these updates

dzver commented:
LGTM

There are failing e2e tests so if nothing of that is related, it's good to go IMO.

creativecoder commented:
Thanks! Re: the Jetpack e2e tests, I think there's something wrong with the circleci/github integration for those tests. If you go to the actually test run page for this branch, they are all passing: https://circleci.com/gh/Automattic/wp-calypso/102125

creativecoder commented:
Looks like there are some issues with the Jetpack e2e test setup, so I've removed that label (p15344253880001000-slack-calypso). I think we can disregard the "failing" tests.

mattsherman commented:
Fixes #26074
